src/ModuleLoaderInterface.php

```diff
@@ -0,0 +1,29 @@
+<?php
+
+namespace Dhii\Modular;
+
+use Traversable;
+
+/**
+ * Something that can load modules.
+ *
+ * Loading should ideally be mostly determined by the module type.
+ *
+ * Since a module is a generic concept, its use can vary from as simple as representing a file to be loaded to a
+ * complete sub-system that requires a substantial amount of initialisation.
+ *
+ * As such, implementors are encouraged to respect the different module types that can exist and load them accordingly.
+ *
+ * @since [*next-version*]
+ */
+interface ModuleLoaderInterface
+{
+    /**
+     * Loads the given module or collection of modules.
+     *
+     * @since [*next-version*]
+     *
+     * @param array|Traversable|ModuleInterface $modules A module or a collection of modules. Default: array()
+     */
+    public function load($modules = array());
```

Comment:
If it only loads a single `ModuleInterface` and it documents that it specifically does so, consumers can simply iterate as needed over the results of the locator.

However I think that this would shift the responsibility of module load order to the consumer. If this is what you meant by "re-arrange modules" then yes, I think the loader should be able to do so. It allows consuming systems to utilize different loaders for different load order strategies.

I have no objections to tightening it to `ModuleInterface[]`. It might also be helpful to union it with `\Traversable` in that case.